### PR TITLE
Jamf v2

### PR DESCRIPTION
**Pull request by barchen1**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/36155

## Description
Jamf v2 integration
includes: 39 commands:

- jamf-get-computers
- jamf-get-computers-basic-subset
- jamf-get-computer-by-id
- jamf-get-computer-by-match 
- jamf-get-computer-general-subset
- jamf-get-computer-location-subset
- jamf-get-computer-purchasing-subset 
- jamf-get-computer-peripherals-subset 
- jamf-get-computer-hardware-subset
- jamf-get-computer-certificates-subset 
- jamf-get-computer-security-subset 
- jamf-get-computer-software-subset
- jamf-get-computer-extension-attributes-subset
- jamf-get-computer-groups-accounts-subset 
- jamf-get-computer-iphones-subset
- jamf-get-computer-configuration-profiles-subset 
- jamf-computer-lock
- jamf-computer-erase
- jamf-get-users
- jamf-get-user-by-id 
- jamf-get-user-by-name
- jamf-get-user-by-email 
- jamf-get-mobile-devices 
- jamf-get-mobile-device-by-id
- jamf-get-mobile-device-by-match 
- jamf-get-mobile-device-general-subset
- jamf-get-mobile-device-location-subset 
- jamf-get-mobile-device-purchasing-subset 
- jamf-get-mobile-device-applications-subset 
- jamf-get-mobile-device-security-subset 
- jamf-get-mobile-device-network-subset 
- jamf-get-mobile-device-certificates-subset 
- jamf-get-mobile-device-extension-attributes-subset
- jamf-get-mobile-device-provisioning-profiles-subset
- jamf-get-mobile-device-groups-subset
- jamf-get-mobile-device-configuration-profiles-subset 
- jamf-get-computers-by-application 
- jamf-mobile-device-lost-mode - beta command
- jamf-mobile-device-erase -beta command
test-module
TPB
UTs
README

note:
The following commands were not tested (as we don't data for it):
 **jamf-mobile-device-lost-mode**
 **jamf-mobile-device-erase**

## Minimum version of Cortex XSOAR
- [x] 5.5.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 1]
```diff
@@ -0,0 +1,982 @@
...
+        return raw_action
+
+
+''' HELPER FUNCTIONS '''
+
+
+def pagination(response, limit, page):
+    if limit > MAX_PAGE_SIZE:
+        limit = MAX_PAGE_SIZE
+    return response[page * limit:(page + 1) * limit]
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:307]: This should be split into several variables with readable names, it's too complex and bug-prone

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+        raw_action = json.loads(xml2json(res.content))
+        return raw_action
+
+
+''' HELPER FUNCTIONS '''
+
+
+def pagination(response, limit, page):
+    if limit > MAX_PAGE_SIZE:
+        limit = MAX_PAGE_SIZE
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:306]: let's make this a one line if

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 3]
```diff
@@ -0,0 +1,982 @@
...
+            'Purchasing Contact': purchasing_computer_response.get('purchasing_contact')
+        }
+
+    elif subset == 'Peripherals':
+        readable_output = computers_response.get('peripherals')
+
+    elif subset == 'Hardware':
+        devices_sizes = []
+        for device in computers_response.get('hardware').get('storage'):
+            devices_sizes.append({device.get('disk'): device.get('size')})
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:385]: can you explain this logic?

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 4]
```diff
@@ -0,0 +1,982 @@
...
+                'Name': user.get('name'),
+
+            })
+    return readable_output
+
+
+def get_mobile_devices_readable_output(response, mobile_id):
+    readable_output = []
+    if mobile_id:
+        mobile_dev_response = response.get('mobile_device').get('general')
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:475]: What does `mobile_dev_response` stand for?

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+
+
+def get_computers_by_app_readable_output(response):
+    readable_output = []
+    versions_list = response.get('versions')
+    for version in versions_list:
+        computers_len = len(version.get('computers'))
+        readable_output.append({
+            'version': version.get('number'),
+            'Sum of computers': computers_len
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:642]: The word `sum` does not fit here IMO.
Perhaps `Total computers` or something like that

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 6]
```diff
@@ -0,0 +1,982 @@
...
+            'Number of groups': len(mobile_device_groups_mobile_response)
+        }
+    return readable_output
+
+
+def get_computers_by_app_readable_output(response):
+    readable_output = []
+    versions_list = response.get('versions')
+    for version in versions_list:
+        computers_len = len(version.get('computers'))
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:639]: IMO the word `len` should not be in this name

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 7]
```diff
@@ -0,0 +1,982 @@
...
+    :type client: ``Client``
+    :param Client: client to use
+
+    :return: 'ok' if test passed, anything else will fail the test.
+    :rtype: ``str``
+    """
+
+    message: str = ''
+    try:
+        if client.get_computers_request():
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:677]: Don't we want to check that something returns?

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 8]
```diff
@@ -0,0 +1,982 @@
...
+
+def get_users_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    user_id = args.get('id')
+    name = args.get('name')
+    email = args.get('email')
+    limit = arg_to_number(args.get('limit', 50))
+    page = arg_to_number(args.get('page', 0))
+
+    user_response = client.get_users_request(user_id, name, email)
+    if not user_id and not name and not email:
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:793]: You can use `any()`

File: Packs/jamf/Integrations/jamf_v2/jamf_v2.py [hunk 9]
```diff
@@ -0,0 +1,982 @@
...
+                                 resp_type='response', error_handler=self._mobile_lost_erase_error_handler)
+
+        raw_action = json.loads(xml2json(res.content))
+        return raw_action
+
+
+''' HELPER FUNCTIONS '''
+
+
+def pagination(response, limit, page):
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamf_v2/jamf_v2.py:304]: Out of all the functions, in my opinion this one deserves a docstring the most.
In general, helper functions need docstrings more than command functions do

**Review by guykeller — CHANGES_REQUESTED** [review]

**Comment by guykeller** [comment]
Great job!!

Some general notes:
1. Some of my notes apply generally to all the code - try to implement them everywhere
2. Naming, descriptions, etc are important IMO, let's give them special attention
3. The builds seem to be failing

File: Packs/jamf/Integrations/jamf_v2/README.md [hunk 10]
```diff
@@ -0,0 +1,1004 @@
+EMM for apple devices (Mac, iPhone, Apple TV, iPad). Can be used to control various configurations via different policies, Install/Uninstall applications, lock devices, smart groups searches and more.
+This integration was integrated and tested with version 10.28.0-t1615386406 of jamf v2
```
[the remark this answers is not in the document]
Comment (tomer-pan) [reply Packs/jamf/Integrations/jamf_v2/README.md]: @barchen1 I think we can change it to:
`This integration was integrated and tested with version 10.28.0 of Jamf Pro`

File: Packs/jamf/Integrations/jamf_v2/jamf_v2_description.md [hunk 11]
```diff
@@ -0,0 +1,7 @@
+
+#### Server URL:
+Please add your server URL (e.g: https://yourdomain.jamfcloud.com)
+#### Username:
+The username of a user with relevant permissions to your Jamf Pro admin console
+#### Password:
+The login password of the above username
```
[the remark this answers is not in the document]
Comment (tomer-pan) [reply Packs/jamf/Integrations/jamf_v2/jamf_v2_description.md]: Nothing else is needed.

File: Packs/jamf/Integrations/integration-jamf.yml [hunk 12]
```diff
@@ -3,10 +3,10 @@ commonfields:
...
-description: Jamf device management. Please note, this integration is currently only compatible with the Jamf V1 API.
+description: Deprecated. Use Jamf v2 instead. Jamf device management. Please note, this integration is currently only compatible with the Jamf V1 API.
```
[the remark this answers is not in the document]
Comment (guykeller) [reply Packs/jamf/Integrations/integration-jamf.yml:9]: Yes

File: Packs/jamf/Integrations/jamfV2/jamfV2.py [hunk 13]
```diff
@@ -0,0 +1,1366 @@
...
+
+    else:
+        command_results.append(CommandResults(
+            readable_output="No endpoints were found",
+            raw_response=outputs,
+        ))
+    return command_results
+
+
+def paging_outputs_dict(total_results, page_size, current_page):
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamfV2/jamfV2.py:777]: This function's name could be improved in my opinion

File: Packs/jamf/Integrations/jamfV2/jamfV2.py [hunk 14]
```diff
@@ -0,0 +1,1366 @@
...
+
+        json_response = json.loads(xml2json(res.content))
+        return json_response
+
+
+''' HELPER FUNCTIONS '''
+
+
+def pagination(response, limit, page):
+    """
```
Comment (guykeller) [inline Packs/jamf/Integrations/jamfV2/jamfV2.py:302]: Let's add a 1-2 lines explanation for what's going on here

**Review by guykeller — APPROVED** [review]
Do one final swipe on readability. Overall looks good

**Comment by barchen1** [comment]
@ShirleyDenkberg Can you please review the PR docs ?

**Comment by ShirleyDenkberg** [comment]
@tomer-pan @guykeller Doc review completed.  @barchen1 Integration README needs to be regenerated.

**Review by guykeller — APPROVED** [review]

**Pull request merged**